### PR TITLE
Driver for Scaleway

### PR DESCRIPTION
**Pull request by nlamirault**
First Pull Request :
https://github.com/docker/machine/pull/947

Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

### CODE REVIEW

**Comment by shouze** [comment]
:+1:

File: drivers/scaleway/scaleway.go [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+	}
+	return nil
+}
+
+func (d *Driver) publicSSHKeyPath() string {
+	return d.GetSSHKeyPath() + ".pub"
+}
+
+func (d *Driver) getClient() *api.ScalewayClient {
+	return api.NewClient(d.UserId, d.Token, d.Organization)
```
Comment (moul) [inline drivers/scaleway/scaleway.go]: The order of arguments does not match https://github.com/nlamirault/go-scaleway/blob/master/api/client.go#L48

File: drivers/scaleway/scaleway.go [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+}
+
+func GetCreateFlags() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:   "scaleway-userid",
+			Usage:  "Scaleway UserID",
+			EnvVar: "SCALEWAY_USERID",
+		},
+		cli.StringFlag{
```
Comment (moul) [inline drivers/scaleway/scaleway.go]: You should remove this optional argument which will always be unique for a token

Use this always: https://github.com/nlamirault/go-scaleway/blob/master/api/client.go#L76

File: drivers/scaleway/scaleway.go [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+	d.waitForServerState(state.Stopped)
+	return nil
+}
+
+func (d *Driver) Remove() error {
+	log.Infof("[Scaleway] Removing instance... ")
+	client := d.getClient()
+	if err := client.DeleteServer(d.Id); err != nil {
+		return err
+	}
```
Comment (moul) [inline drivers/scaleway/scaleway.go]: You should delete the dedicated ssh key here too

File: drivers/scaleway/scaleway.go [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+		cli.StringFlag{
+			Name:   "scaleway-organization",
+			Usage:  "Organization identifier",
+			EnvVar: "SCALEWAY_ORGANIZATION",
+		},
+		cli.StringFlag{
+			Name:   "scaleway-image",
+			Usage:  "Image identifier",
+			EnvVar: "SCALEWAY_IMAGE",
+		},
```
Comment (moul) [inline drivers/scaleway/scaleway.go]: You should compute the image ID dynamically if this option is left empty

File: drivers/scaleway/scaleway.go [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+		cli.StringFlag{
+			Name:   "scaleway-volumes",
+			Usage:  "Volumes identifier",
+			EnvVar: "SCALEWAY_VOLUMES",
```
Comment (moul) [inline drivers/scaleway/scaleway.go]: How this option works ?

File: drivers/scaleway/scaleway.go (hunk 1 above)
Comment (nlamirault) [inline drivers/scaleway/scaleway.go]: done. Thanks

**Comment by macnibblet** [comment]
Any updates on this ?

**Comment by Spy-Seth** [comment]
:+1:

**Comment by lfalorni** [comment]
:+1:

**Comment by jeremyFreeAgent** [comment]
:+1:

**Comment by ehazlett** [comment]
Hi , thanks for your efforts and persistence in submitting this driver. We are extremely excited that there is so much interest in Docker Machine and we really appreciate your interest. However, at this time it is proving to be extremely difficult for us to keep up with reviewing and testing each of these drivers for inclusion in the Machine core. We really want to switch to a more pluggable model, as well as polish up a few things about the driver model which need to be changed to ensure a smooth and sustainable future.

Therefore, we will be moving to a plugin model for 0.5 and would love to have you involved in the design and development process. We are closing the outstanding driver PRs at this time, but please keep the code. We will stick closely to the current driver interface and you should be able to re-use a lot (if not all) of the existing driver along with the new plugin model. We will be moving all of the drivers which are merged directly into Machine today to the plugin model when it is available, so there will be no special treatment of those, and there will be documentation outlining the process of developing and using a Docker Machine driver plugin.

With all of that being said, we want to apologize for the lack of feedback on your pull request. As contributors ourselves, we understand that being left in limbo is no fun. We would have liked to address this sooner, and in the future we will be more responsive around these kinds of issues.

Once again, we thank you for the contribution and the tremendous support. Keep hacking strong!

If you want to contribute to the design of the plugin model, we'd love to get your input on this issue where we will be planning it:

https://github.com/docker/machine/issues/1626

@ehazlett & @nathanleclaire

**Pull request closed without merging**